### ISSUE TITLE
perl print_long_double test fails on ppc64le with  -mabi=ieeelongdouble

### ISSUE THREAD

**Issue by tstellar** [llvm/llvm-project#61913]
I spent some time investigating this and my guess is that it is some kind of issue with the calling convention used for IEEE long double.  The failing test case passes these values through several `va_args` functions.

I observed this bug in both Clang 15 and Clang 16.  Here is how to reproduce:

```shell
dnf install systemtap-sdt-devel bzip2-devel compat-gdbm-devel git

git clone https://github.com/Perl/perl5
cd perl5
git checkout v5.36.0

/bin/sh Configure -Dusedevel -des -Doptimize=none '-Dccflags=-O2  -mabi=ieeelongdouble -g -Wp,-D_FORTIFY_SOURCE=2 -m64 -mcpu=power8 -mtune=power8'  '-Dccdlflags=-Wl,--enable-new-dtags -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now    -Wl,--build-id=sha1 ' '-Dlddlflags=-shared -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now  -Wl,--build-id=sha1 ' -Dshrpdir=/usr/lib64 -DDEBUGGING=-g -Dversion=5.36.0 -Dmyhostname=localhost -Dcc=clang '-Dcf_by=Red Hat, Inc.' -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl5/5.36 -Dsitearch=/usr/local/lib64/perl5/5.36 -Dprivlib=/usr/share/perl5 -Dvendorlib=/usr/share/perl5/vendor_perl -Darchlib=/usr/lib64/perl5 -Dvendorarch=/usr/lib64/perl5/vendor_perl -Darchname=ppc64le-linux-thread-multi '-Dlibpth=/usr/local/lib64 /lib64 /usr/lib64' -Duseshrplib -Dusethreads -Duseithreads -Dusedtrace=/usr/bin/dtrace -Duselargefiles -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly '-Dpager=/usr/bin/less -isr' -Dd_gethostent_r_proto -Ud_endhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin -Dusesitecustomize -Duse64bitint

# Build
make -j8

# Run test
LD_LIBRARY_PATH=.:${LD_LIBRARY_PATH} ./perl -I./lib ext/XS-APItest/t/printf.t
```

**Comment by tuliom** [comment]
I suspect the issue is actually in the post-processed C source code.
Looking at the libperl.so generated with GCC, all references to `*printf*` are going to ieee128 symbols, which is what I expected to happen in an IEEE 128-bit long double case.

However, when compiling with clang, part of the calls go to IBM long double symbols (without the `ieee128` suffix), e.g.:

```
    17: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND __vsnprintf_chk@GLIBC_2.17 (2)       [<localentry>: 8]
   139: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND __vfprintf_chkieee128@GLIBC_2.32 (8)         [<localentry>: 8]
```

These symbols are chosen by the C standard library, which is glibc in this case.
glibc choose them based on C macros provided by the compiler and user code (i.e. perl).

**Comment by tuliom** [comment]
This appears to be a glibc bug with `_FORTIFY_SOURCE`.
I created a reproducer:

```
$ cat test-vsnprintf.c 
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>

char *
foo(const char *fmt, ...)
{
    size_t size = 0;
    char *p = NULL;
    va_list ap;

    /* Determine required size. */

    va_start(ap, fmt);
    vsnprintf(p, size, fmt, ap);
    va_end(ap);

    return p;
}
$ clang -c -mabi=ieeelongdouble -O1 -Wp,-D_FORTIFY_SOURCE=2 test-vsnprintf.c
$ readelf -aW test-vsnprintf.o | grep -E 'UND.*printfieee128'
$
```

This is the expected output:
```
$ clang -c -mabi=ieeelongdouble -O1 test-vsnprintf.c
$ readelf -aW test-vsnprintf.o | grep -E 'UND.*printfieee128'
     5: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __vsnprintfieee128
$
```

**Comment by tuliom** [comment]
I've just confirmed the problem is happening because clang is ignoring an asm redirection.
I reduced the code to the following reproducer:

```
typedef long unsigned int size_t;
typedef __builtin_va_list __gnuc_va_list;
typedef __gnuc_va_list va_list;


extern int vsnprintf (char *__restrict __s, size_t __maxlen,
        const char *__restrict __format, __gnuc_va_list __arg)
     __attribute__ ((__nothrow__)) __attribute__ ((__format__ (__printf__, 3, 0)));

extern __typeof (vsnprintf) vsnprintf __asm("vsnprintfieee128");


extern __inline __attribute__ ((__always_inline__)) __attribute__ ((__gnu_inline__)) __attribute__ ((__artificial__)) int
__attribute__ ((__nothrow__ )) vsnprintf (char *__restrict __s, size_t __n, const char *__restrict __fmt, __gnuc_va_list __ap)

{
  return __builtin___vsnprintf_chk (__s, __n, 1 - 1,
        __builtin_object_size (__s, 1 > 1), __fmt, __ap);
}

char *
foo(const char *fmt, ...)
{
    size_t size = 0;
    char *p = ((void*)0);
    va_list ap;

    __builtin_va_start(ap, fmt);
    vsnprintf(p, size, fmt, ap);
    __builtin_va_end(ap);

    return p;
}
```

As with the previous reproducer, the call is expected to go to `vsnprintfieee128`, but clang ends up calling `vsnprintf`:

```
$ clang -c -mabi=ieeelongdouble -O1 test-redir.c 
$ readelf -aW test-redir.o | grep printf
000000000000004c  000000050000000a R_PPC64_REL24          0000000000000000 vsnprintf + 0
     5: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND vsnprintf
```

**Comment by tstellar** [comment]
@serge-sans-paille Have you seen anything like this during your _FORTIFY_SOURCE work?

**Comment by tbaederr** [comment]
> I've just confirmed the problem is happening because clang is ignoring an asm redirection.
> I reduced the code to the following reproducer:

Clang sees the `__asm` label and adds it to the functiondecl. It just ends up calling the third and last `vsnprintf` declared in the reproducer.

Removing the `always_inline` attribute from that last declaration fixes the reproducer for me.

**Comment by tbaederr** [comment]
If the `always_inline` attribute is applied, we do not treat a call to `vsnprintf` as a builtin and end up calling `EmitCall` from `CodeGenFunction::EmitCallExpr`, while without it, we treat it as a builtin and call `EmitBuiltinExpr`. That's because `OnlyHasInlineBuiltinDecaration()` returns `false` now.

I'm not sure what's the right thing to do here though. Should it still be treated as a builtin expr? CC @rjmccall / @efriedma-quic for the codegen parts.

**Comment by efriedma-quic** [comment]
It looks like in gcc, renaming the declaration of __vsnprintf_chk via "asm" also affects calls to __builtin___vsnprintf_chk.  That doesn't work in clang; we don't lookup the name in the original AST when we generate __builtin_ calls, and we resolve "asm" much earlier (in clang IR generation).

I see two possible approaches here:

- Keep some sort of "asm" mapping side-table, and use it to compute the names of builtin calls in clang, and encode it into the IR so TargetLibraryInfo can understand the correct names.
- Teach clang/LLVM about all the C library functions that get renamed when -mabi=ieeelongdouble is passed; essentially, continue to ignore the "asm" markings in the context of __builtin_ calls, but get the names right anyway because we know what they should be.

The latter is sort of hacky, but probably significantly easier...

**Comment by tbaederr** [comment]
Isn't the latter what we're already doing here:
 https://github.com/llvm/llvm-project/blob/aba32abe2d93133a46f67764937a1c85d455dce8/clang/lib/CodeGen/CGBuiltin.cpp#L102-L110

**Comment by ecnelises** [comment]
Yes, clang already mutates some builtin calls if ABI is ieeelongdouble (to be consistent with GCC).

**Comment by tbaederr** [comment]
That's the change I added locally as well. I see you initially added that list, so if you think that's the right change to make, I can prepare a patch.

**Comment by tuliom** [comment]
@tbaederr Keep in mind there are 3 functions affected. I reduced the reproducer to a single function in order to simplify things, but the following functions have the same issue in perl:
```
__vsnprintf_chk
__sprintf_chk
__snprintf_chk
```

**Comment by ecnelises** [comment]
Maybe we need more changes than that? The original list is incomplete, and some `*printf`/`*scanf` GCC builtins are not available in Clang. I will post a patch to complete it.

**Comment by ecnelises** [comment]
Created https://reviews.llvm.org/D150087 .

**Comment by ecnelises** [comment]
Landed D150087, the reproducer should emit call to symbol with suffix `ieee128` (actually `grep -E 'UND.*printf_chkieee128'`, not `grep -E 'UND.*printfieee128'`).
